hazelcast-jet-core/src/main/java/com/hazelcast/jet/impl/execution/BlockingProcessorTasklet.java

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2008-2017, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.jet.impl.execution;
+
+import com.hazelcast.jet.Outbox;
+import com.hazelcast.jet.Processor;
+import com.hazelcast.jet.Processor.Context;
+import com.hazelcast.jet.impl.util.ProgressState;
+import com.hazelcast.util.Preconditions;
+
+import javax.annotation.Nonnull;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+import static com.hazelcast.jet.impl.execution.ExecutionService.IDLER;
+import static com.hazelcast.jet.impl.util.DoneItem.DONE_ITEM;
+
+/**
+ * Tasklet that drives a non-cooperative processor.
+ */
+public class BlockingProcessorTasklet extends ProcessorTaskletBase {
+
+    private final BlockingOutbox outbox;
+    private CompletableFuture<?> jobFuture;
+
+    public BlockingProcessorTasklet(
+            String vertexName, Context context, Processor processor,
+            List<InboundEdgeStream> instreams, List<OutboundEdgeStream> outstreams
+    ) {
+        super(vertexName, context, processor, instreams, outstreams);
+        Preconditions.checkFalse(processor.isCooperative(), "Processor is cooperative");
+        outbox = new BlockingOutbox();
+        processor.init(outbox, context);
+    }
+
+    @Override
+    public final boolean isCooperative() {
+        return false;
+    }
+
+    @Override
+    public void init(CompletableFuture<?> jobFuture) {
+        this.jobFuture = jobFuture;
+    }
+
+    @Override @Nonnull
+    public ProgressState call() {
+        try {
+            progTracker.reset();
+            tryFillInbox();
+            if (progTracker.isDone()) {
+                complete();
+            } else if (!inbox().isEmpty()) {
+                process();
+            }
+            return progTracker.toProgressState();
+        } catch (JobFutureCancelled e) {
+            return ProgressState.DONE;
+        }
+    }
+
+    private void process() {
+        progTracker.madeProgress();
+        processor.process(currInstream.ordinal(), inbox());
+    }
+
+    private void complete() {
+        progTracker.madeProgress();
+        if (!processor.complete()) {
+            progTracker.notDone();
+            return;
+        }
+        for (OutboundEdgeStream outstream : outstreams) {
+            outbox.add(outstream.ordinal(), DONE_ITEM);
+        }
+    }
```

Comment:
The method could be just: 

    private void complete() {
        progTracker.madeProgress();
        if (processor.complete()) {
            outbox.add(DONE_ITEM);
        } else {
            progTracker.notDone();
        }
    }